### PR TITLE
Added linear constraint check and updated documentation

### PR DESCRIPTION
**Pull request by kanekosh**
## Purpose

pyOptSparse computes the linear constraints exclusively based on `jac`, `lower`, and `upper` entries of `addConGroup`, and it ignores the linear constraint value returned by the user-defined obj/con function. This is confusing and can result in unexpected behavior (see related #296).

This PR adds a linear constraint checkerto make sure that a user-defined obj/con function does **not** return linear constraint values. If it does, pyOptSparse will raise an error. I also updated the documentation accordingly.

I fixed a few existing tests that have been computing linear constraint values in the obj/con function (which we should not). I also added a test to check if pyOptSparse raises an error as expected.

This change will break some existing runscripts that return linear constraint values in the user-defined function. OpenMDAO's pyOptSparseDriver will be fine. Not sure about MACH runscripts (haven't checked).

## Type of change

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [x] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Checklist

- [ ] I have run `flake8` and `black` to make sure the Python code adheres to PEP-8 and is consistently formatted
- [ ] I have formatted the Fortran code with `fprettify` or C/C++ code with `clang-format` as applicable
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

### CODE REVIEW

File: tests/test_snopt_bugfix.py [hunk 1]
```diff
@@ -21,35 +21,6 @@ def objfunc(xdict):
...
 
     funcs["obj"] = (x - 3.0) ** 2 + x * y + (y + 4.0) ** 2 - 3.0
-    conval = -x + y
-    funcs["con"] = conval
-
-    fail = False
-    return funcs, fail
-
-
-def objfunc_no_con(xdict):
```
Comment (marcomangano) [inline tests/test_snopt_bugfix.py:31]: Nice catch, I assume the new check had these tests failing right away, right?

**Review by marcomangano — APPROVED** [review]
No additional comments on the check we came up with, thanks for fixing the tests!!

File: tests/test_snopt_bugfix.py (hunk 1 above)
Comment (kanekosh) [reply tests/test_snopt_bugfix.py:31]: Yes you're right

**Comment by marcomangano** [comment]
@lamkina @ewu63 @sseraj can you check and merge this soon so we can go ahead and make a new release?

**Comment by lamkina** [comment]
Do we have a unit test that confirms the standard way of adding linear constraints is working as expected? If not, can that be added to this PR? If it exists and I missed it, then point me to the test and I'll approve this PR.

This would just be as simple as adding a few linear constraints to a valid opt prob and then making sure the number of linear constraints and the Jacobian are correct.

**Comment by kanekosh** [comment]
I don't know if there's a test that checks the number of linear constraints, but we already have a few tests that include linear constraints and assert equal on the optimal solution, which validates that linear constraints are properly imposed.

**Comment by lamkina** [comment]
That's good enough for me. Thanks for the double check!

**Review by lamkina — APPROVED** [review]

**Pull request merged**